### DISCUSSION TITLE
I need guidance for transforming OpenCL code to RISC-V assembly form

### DISCUSSION THREAD

**Discussion by suleymanemre** [pocl/pocl#1467] [Q&A]
Hello, I am very new to this subject, so I need your guidance. Thank you in advance to everyone who takes the time.
I would like to start by summarizing the project I want to do as follows. We want to run OpenCL on a RISC-V core of our own design. While doing this, we thought we should use POCL. We took such a path because of the article you can find at the link here https://carrv.github.io/2021/papers/CARRV2021_paper_36_Blaise.pdf.

At this point, I have installed POCL on my computer, but I could not run any sample code. The $clinfo command shows that there is no opencl platform on my system.

What should I do at this point? The point I want to achieve is to properly convert OpenCL Host Code and Kernel Code into RISC-V assembly format.

**Comment by pjaaskel** [comment]
Hi. Which platform you are building in? We have basic RISC-V support working upstream, but if you want to go for the "standalone" compilation mode they discuss in that paper, you best ask the authors of the paper you linked.

These are some of the relevant parts in the user manual that should be helpful: http://portablecl.org/docs/html/install.html#risc-v-support http://portablecl.org/docs/html/development.html#using-pocl-from-the-build-tree.

**Reply by suleymanemre** [reply]
I am developing on both windows and ubuntu. I read that part already, thank you for your response.

To explain my problem more clearly. I can say that I have a core that supports the risc-v instruction set that I designed, and I want to load the codes that I have converted from OpenCL into the construction memory on this core. If I can do this, I will know that my system supports OpenCL.

Who did the tests mentioned on POCL's page? If I ask him, maybe I can create my own testing process. Also I will try to ask authors of paper. Thanks again.

**Comment by jansol** [comment]
> I have installed POCL on my computer, but I could not run any sample code. The $clinfo command shows that there is no opencl platform on my system.

Does this mean that you installed PoCL from your package manager? When using PoCL in a standard PC environment it's generally assumed that you have an ICD loader installed and the default build is set up so that said loader will find the PoCL platform. Make sure you have the `ocl-icd-libopencl1` package (ubuntu/debian - other distros may have a slightly different name; alternatively the khronos opencl icd loader package) installed.

If you built PoCL from source, note that clinfo is linked against the system `libOpenCL.so` so you need to either ensure that is an ICD loader (simply installing an ICD loader package solves this) or that clinfo will find the PoCL `libOpenCL.so` before looking at the system one. (e.g. by pointing `LD_LIBRARY_PATH` at your build directory to have the system look there before it checks the usual system paths)

**Reply by suleymanemre** [reply]
While installing, I followed the steps in the download manual in order. In windows now i can see my AMD GPU but in ubuntu (in other computer) i cannot see my GPU its about my Graphic card driver (its another problem dont worry about it). 

I need to understand how to define another device to pocl.

**Reply by jansol** [reply]
If you compiled PoCL from sources yourself the ICD loader won't find it (it only looks at certain system paths for OpenCL drivers by default). You need to either install PoCL system-wide to put the `.icd` file in the right place OR point the `OCL_ICD_VENDORS` environment varialbe to the generated `.icd` file as mentioned in http://portablecl.org/docs/html/using.html#installable-client-driver-icd (the latter tells the loader to ignore system paths and use your development build of PoCL. Note that you'll also need to set `POCL_BUILDING=1` if you want to use PoCL directly from its build directory)

**Comment by suleymanemre** [comment]
Hello, I found a way for this. There is a compiler for Ventus GPGPU in the repo in the [link](https://github.com/THU-DSP-LAB/llvm-project)  I added. After the installation, I was able to convert and print using pocl. It looks ideal to try. Thank you all for your answers.
